File: services/compact/pendingCompactRequest.ts
/**
 * W48 — Pending compact request buffer.
 *
 * Single-slot buffer for stream-json compact_conversation control requests.
 * The control_request handler enqueues here; the query loop safe point
 * dequeues and executes. No execution happens outside the query loop.
 *
 * Invariants:
 * - Single slot: enqueue returns false if a request is already pending.
 * - No ToolUseContext / CacheSafeParams construction here.
 * - No compactConversation calls here.
 */

export type PendingCompactRequest = {
  requestId: string
  mode: 'manual'
  dryRun: boolean
  customInstructions?: string
  enqueuedAt: number
}

let pendingRequest: PendingCompactRequest | null = null

/** Timeout for queued compact requests (ms). */
export const COMPACT_REQUEST_TIMEOUT_MS = 60_000

/**
 * Enqueue a compact request. Returns false if a request is already pending.
 */
export function enqueuePendingCompactRequest(
  req: Omit<PendingCompactRequest, 'enqueuedAt'>,
): { ok: true } | { ok: false; reason: string } {
  if (pendingRequest !== null) {
    return {
      ok: false,
      reason: 'another compact request is already pending',
    }
  }
  pendingRequest = { ...req, enqueuedAt: Date.now() }
  return { ok: true }
}

/**
 * Dequeue the pending request. Returns null if none pending.
 * Checks for timeout — if timed out, returns null and the caller
 * should emit a compact_completed(failed) event.
 */
export function dequeuePendingCompactRequest(): PendingCompactRequest | null {
  if (pendingRequest === null) return null
  if (Date.now() - pendingRequest.enqueuedAt > COMPACT_REQUEST_TIMEOUT_MS) {
    const timedOut = pendingRequest
    pendingRequest = null
    return timedOut // caller checks timeout via hasCompactRequestTimedOut
  }
  const req = pendingRequest
  pendingRequest = null
  return req
}

/**
 * Peek at the pending request without dequeuing.
 */
export function getPendingCompactRequest(): PendingCompactRequest | null {
  return pendingRequest
}

/**
 * Check whether a pending request exists.
 */
export function hasPendingCompactRequest(): boolean {
  return pendingRequest !== null
}

/**
 * Check whether the pending request has timed out.
 */
export function hasCompactRequestTimedOut(): boolean {
  if (pendingRequest === null) return false
  return Date.now() - pendingRequest.enqueuedAt > COMPACT_REQUEST_TIMEOUT_MS
}

/**
 * Clear the pending request unconditionally.
 */
export function clearPendingCompactRequest(): void {
  pendingRequest = null
}

File: crates/mossen-agent/src/services/compact/pending_compact_request.rs
//! Pending compact request buffer.
//!
//! Single-slot buffer for stream-json compact_conversation control requests.
//! The control_request handler enqueues here; the query loop safe point
//! dequeues and executes.

use std::sync::Mutex;
use std::time::{Duration, Instant};

/// Timeout for queued compact requests.
pub const COMPACT_REQUEST_TIMEOUT: Duration = Duration::from_secs(60);

/// A pending compact request.
#[derive(Debug, Clone)]
pub struct PendingCompactRequest {
    pub request_id: String,
    pub mode: CompactMode,
    pub dry_run: bool,
    pub custom_instructions: Option<String>,
    pub enqueued_at: Instant,
}

/// Compact mode (currently only manual).
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CompactMode {
    Manual,
}

/// Global pending request slot.
static PENDING_REQUEST: Mutex<Option<PendingCompactRequest>> = Mutex::new(None);

/// Enqueue a compact request. Returns Err if a request is already pending.
pub fn enqueue_pending_compact_request(
    request_id: String,
    mode: CompactMode,
    dry_run: bool,
    custom_instructions: Option<String>,
) -> Result<(), String> {
    let mut slot = PENDING_REQUEST.lock().unwrap();
    if slot.is_some() {
        return Err("another compact request is already pending".to_string());
    }
    *slot = Some(PendingCompactRequest {
        request_id,
        mode,
        dry_run,
        custom_instructions,
        enqueued_at: Instant::now(),
    });
    Ok(())
}

/// Dequeue the pending request. Returns None if none pending.
/// If timed out, still returns the request (caller checks via `has_compact_request_timed_out`).
pub fn dequeue_pending_compact_request() -> Option<PendingCompactRequest> {
    let mut slot = PENDING_REQUEST.lock().unwrap();
    slot.take()
}

/// Peek at the pending request without dequeuing.
pub fn get_pending_compact_request() -> Option<PendingCompactRequest> {
    let slot = PENDING_REQUEST.lock().unwrap();
    slot.clone()
}

/// Check whether a pending request exists.
pub fn has_pending_compact_request() -> bool {
    let slot = PENDING_REQUEST.lock().unwrap();
    slot.is_some()
}

/// Check whether the pending request has timed out.
pub fn has_compact_request_timed_out() -> bool {
    let slot = PENDING_REQUEST.lock().unwrap();
    match &*slot {
        Some(req) => req.enqueued_at.elapsed() > COMPACT_REQUEST_TIMEOUT,
        None => false,
    }
}

/// Clear the pending request unconditionally.
pub fn clear_pending_compact_request() {
    let mut slot = PENDING_REQUEST.lock().unwrap();
    *slot = None;
}
